src/showcase/nz-demo-form/nz-demo-form-dynamic.component.ts

```diff
@@ -53,32 +62,39 @@ import {
 export class NzDemoFormDynamicComponent implements OnInit {
   validateForm: FormGroup;
   controlArray = [];
-  index = 0;
 
   addField(e?: MouseEvent) {
     if (e) {
       e.preventDefault();
     }
-    this.controlArray.push(this.index);
-    this.validateForm.addControl(`passenger${this.index}`, new FormControl(null, Validators.required));
-    this.index++;
+    const id = (this.controlArray.length > 0) ? this.controlArray[this.controlArray.length - 1].id + 1 : 0;
+
+    const control = {
+      id,
+      controlInstance: `passenger${id}`
+    };
+    const index = this.controlArray.push(control);
+    console.log(this.controlArray[this.controlArray.length - 1]);
+    this.validateForm.addControl(this.controlArray[index - 1].controlInstance, new FormControl(null, Validators.required));
   }
 
   removeField(i, e: MouseEvent) {
     e.preventDefault();
     if (this.controlArray.length > 1) {
-      this.controlArray.splice(i, 1);
-      this.validateForm.removeControl(`passenger${this.index}`);
+      const index = this.controlArray.indexOf(i);
```

Comment:
This is the key code.
If you don't agree with me, just add these two lines to the original code.